### ISSUE TITLE
[Provider]: Implement exclusive relationship management resources

### ISSUE THREAD

**Issue by jar-b** [hashicorp/terraform-provider-aws#39376]
### Description

This meta issue will track the progress of the effort to implement "exclusive relationship management" resources, as described in [this proposal](https://github.com/hashicorp/terraform-provider-aws/pull/39203) (see the [rendered version](https://hashicorp.github.io/terraform-provider-aws/design-decisions/exclusive-relationship-management-resources/) in the contributor guide). 

IAM inline policies:
- [x] https://github.com/hashicorp/terraform-provider-aws/pull/39203
- [ ] [New Resource]: `aws_iam_user_policies_exclusive`
- [ ] [New Resource]: `aws_iam_group_policies_exclusive`

IAM customer managed policies:
- [ ] [New Resource]: `aws_iam_role_policy_attachments_exclusive`
- [ ] [New Resource]: `aws_iam_user_policy_attachments_exclusive`
- [ ] [New Resource]: `aws_iam_group_policy_attachments_exclusive`

VPC security groups:
- [ ] [New Resource]: `aws_ec2_network_interface_sg_attachments_exclusive`
- [ ] [New Resource]: `aws_vpc_security_group_rules_exclusive` 
  - naming TBD. We could alternatively have ingress and egress variants to match the format of the ingress/egress rule resources.

### References

Relates #39203

### Would you like to implement a fix?

None

**Comment by lorengordon** [comment]
I love this sooooo much! Here's another related feature request:

* https://github.com/hashicorp/terraform-provider-aws/issues/26352

**Comment by lorengordon** [comment]
Commenting here for visibility since the [proposal is already approved and merged](https://github.com/hashicorp/terraform-provider-aws/pull/39204), but what is the expected config migration path for practitioners? For example, if I am currently using `aws_iam_role` with the `inline_policy` argument exactly for the exclusive management feature, how do I move to the new pattern without impacting existing resources?

**Comment by yermulnik** [comment]
> Commenting here for visibility since the [proposal is already approved and merged](https://github.com/hashicorp/terraform-provider-aws/pull/39204), but what is the expected config migration path for practitioners? For example, if I am currently using `aws_iam_role` with the `inline_policy` argument exactly for the exclusive management feature, how do I move to the new pattern without impacting existing resources?

+1
It's been for years that we've been enforcing `inline_policy` blocks inside `aws_iam_role` resource to ensure exclusive inline policies management by TF and now we're to migrate to `aws_iam_role_policy` with auxiliary `aws_iam_role_policies_exclusive` which is to manage at least two standalone resources outside `aws_iam_role` including policies being decoupled from the IAM Role resource to decrease visibility 😢

**Comment by yermulnik** [comment]
I believe there's a good reasoning behind deprecating `inline_policy` in favour of when one looks into `aws_iam_role` resource and sees no inline policies attached because they are managed by at least two other standalone resources and each may be a part of separate `.tf` file 🤷🏻

**Comment by lorengordon** [comment]
> Commenting here for visibility since the [proposal is already approved and merged](https://github.com/hashicorp/terraform-provider-aws/pull/39204), but what is the expected config migration path for practitioners? For example, if I am currently using `aws_iam_role` with the `inline_policy` argument exactly for the exclusive management feature, how do I move to the new pattern without impacting existing resources?

I went ahead and demoed a quick change to our iam-principals module to see what the migration path looks like currently. I think it needs some work. It's going to be a big lift, without any way to `move` a resource created by an `inline_policy` block to the `aws_iam_policy` resource.

PR here: https://github.com/plus3it/terraform-aws-tardigrade-iam-principals/pull/205

To see what I mean:
1. checkout the default branch for that project
2. run init/apply in `tests/create_roles/prereq`
3. run init/apply in `tests/create_roles`
4. checkout the branch `feat/new-exclusive-pattern`
5. run plan in `tests/create_roles`

You'll see that it wants to create the inline role policies _again_.

Of course, a user could write `import` blocks to import all the inline policies, but that's a little nuts. It would be amazing if I could write a `moved` block for the module, where the `from` is actually the `inline_policy` block on the `iam_role` resource, and the `to` is the new `iam_role_policy` resource.

@jar-b Thoughts?

**Comment by lorengordon** [comment]
Also, managed prefix lists and their entries are another resource that could benefit from this new exclusive relationship pattern...

* https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ec2_managed_prefix_list#entry

**Comment by jar-b** [comment]
To migrate an existing inline policy to the standalone resource, an [`import` block](https://developer.hashicorp.com/terraform/language/import) (or alternatively a manual import via the Terraform CLI) is required. Unfortunately a [`moved` block](https://developer.hashicorp.com/terraform/language/moved) cannot be used, as the shape of the two resources (`aws_iam_role` and `aws_iam_role_policy`) do not match.

However, because the `inline_policy` argument is Optional/Computed (meaning removal of this argument will _not_ trigger deletion of inline policy, just change it to track as computed only), this step can be accomplished in a single apply. 


Here's a small example (the configuration is hidden by default to improve readability):

<details>
  <summary>Show/Hide Configuration</summary>

The uncommented configuration below represents the existing state where inline policies are defined directly on the `aws_iam_role` resource.

```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 5.0"
    }
  }
}

# Configure the AWS Provider
provider "aws" {}

data "aws_iam_policy_document" "trust" {
  statement {
    actions = ["sts:AssumeRole"]
    principals {
      type        = "Service"
      identifiers = ["ec2.amazonaws.com"]
    }
  }
}

data "aws_iam_policy_document" "inline" {
  statement {
    actions = ["s3:ListBucket"]
    resources = [
      "arn:aws:s3:::some-bucket-a",
      "arn:aws:s3:::some-bucket-b",
    ]
  }
}

resource "aws_iam_role" "test" {
  name               = "jb-test-inline-policy-migration"
  assume_role_policy = data.aws_iam_policy_document.trust.json

  # Remove this when the new standalone resource is added
  inline_policy {
    name   = "test-inline"
    policy = data.aws_iam_policy_document.inline.json
  }
}

### Start of new resources ###
# # This can be removed after the apply in which the resource is imported
# import {
#   to = aws_iam_role_policy.test
#   id = "jb-test-inline-policy-migration:test-inline"
# }
#
# resource "aws_iam_role_policy" "test" {
#   name   = "test-inline"
#   role   = aws_iam_role.test.name
#   policy = data.aws_iam_policy_document.inline.json
# }
# #
# resource "aws_iam_role_policies_exclusive" "test" {
#   role_name = aws_iam_role.test.name
#   policy_names = [
#     aws_iam_role_policy.test.name,
#   ]
# }
### End of new resources ###
```
</details>

To covert this to use the standalone `aws_iam_role_policy` resource, do the following:

1. Remove the `inline_policy` block.
1. Uncomment the `### Start of new resources ###` section.
1. `terraform apply`. There should be 1 resource to import and 1 to add.

```console
% terraform apply -auto-approve
data.aws_iam_policy_document.inline: Reading...
data.aws_iam_policy_document.trust: Reading...
data.aws_iam_policy_document.inline: Read complete after 0s [id=748695472]
data.aws_iam_policy_document.trust: Read complete after 0s [id=2851119427]
aws_iam_role.test: Refreshing state... [id=jb-test-inline-policy-migration]
aws_iam_role_policy.test: Preparing import... [id=jb-test-inline-policy-migration:test-inline]
aws_iam_role_policy.test: Refreshing state... [id=jb-test-inline-policy-migration:test-inline]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_iam_role_policies_exclusive.test will be created
  + resource "aws_iam_role_policies_exclusive" "test" {
      + policy_names = [
          + "test-inline",
        ]
      + role_name    = "jb-test-inline-policy-migration"
    }

  # aws_iam_role_policy.test will be imported
    resource "aws_iam_role_policy" "test" {
        id          = "jb-test-inline-policy-migration:test-inline"
        name        = "test-inline"
        name_prefix = null
        policy      = jsonencode(
            {
                Statement = [
                    {
                        Action   = "s3:ListBucket"
                        Effect   = "Allow"
                        Resource = [
                            "arn:aws:s3:::some-bucket-b",
                            "arn:aws:s3:::some-bucket-a",
                        ]
                    },
                ]
                Version   = "2012-10-17"
            }
        )
        role        = "jb-test-inline-policy-migration"
    }

Plan: 1 to import, 1 to add, 0 to change, 0 to destroy.
aws_iam_role_policy.test: Importing... [id=jb-test-inline-policy-migration:test-inline]
aws_iam_role_policy.test: Import complete [id=jb-test-inline-policy-migration:test-inline]
aws_iam_role_policies_exclusive.test: Creating...
aws_iam_role_policies_exclusive.test: Creation complete after 1s

Apply complete! Resources: 1 imported, 1 added, 0 changed, 0 destroyed.
```

4. After a successful apply, the `import` block can be removed, and there are no planned changes.

```console
% terraform plan
data.aws_iam_policy_document.trust: Reading...
data.aws_iam_policy_document.inline: Reading...
data.aws_iam_policy_document.trust: Read complete after 0s [id=2851119427]
data.aws_iam_policy_document.inline: Read complete after 0s [id=748695472]
aws_iam_role.test: Refreshing state... [id=jb-test-inline-policy-migration]
aws_iam_role_policy.test: Refreshing state... [id=jb-test-inline-policy-migration:test-inline]
aws_iam_role_policies_exclusive.test: Refreshing state...

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
```

**Comment by jar-b** [comment]
In theory if all previous versions of the module guarantee the inline policies will already exist, you could publish a version containing the import directive in a non-breaking way. However, a breaking major version change may be preferable to avoid cases where users need to sequence minor versions in such a way to guarantee the import conditions are satisfied. 

The `terraform-aws-s3-bucket` module may provide some precedent for this in how it handled breaking arguments which were previously defined inline on the `aws_s3_bucket` resource into their own standalone resources between v2 and v3.
https://github.com/terraform-aws-modules/terraform-aws-s3-bucket/blob/v4.1.2/UPGRADE-3.0.md

**Comment by lorengordon** [comment]
>In theory if all previous versions of the module guarantee the inline policies will already exist, you could publish a version containing the import directive in a non-breaking way.

I must have missed something... When did it become possible to put `import` blocks in reusable modules? I've only ever used `import` blocks as part of root modules...

**Comment by lorengordon** [comment]
>Unfortunately a [moved block](https://developer.hashicorp.com/terraform/language/moved) cannot be used, as the shape of the two resources (aws_iam_role and aws_iam_role_policy) do not match.

Wasn't there a fairly recent change/feature in Terraform core that made it possible to complete a `move` across resources with different shapes? I know you can [migrate from `null_resource` to `terraform_data`](https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource), for example. See:

* https://github.com/hashicorp/terraform/releases/tag/v1.8.0
* https://github.com/hashicorp/terraform-plugin-framework/pull/917.

**Comment by jar-b** [comment]
> I must have missed something... When did it become possible to put `import` blocks in reusable modules? I've only ever used `import` blocks as part of root modules...

🤦 - thanks for the correction. I'm clearly not a module author and will therefore refrain from additional suggestions.

> Wasn't there a fairly recent change/feature in Terraform core that made it possible to complete a move across resources with different shapes?

Yes. My assumption about the schema requirements is indeed out of date. However, we are still limited in our ability to support the `moved` block between these resource types because `aws_iam_role_policy` is implemented with [Terraform Plugin SDK V2](https://developer.hashicorp.com/terraform/plugin/sdkv2), which does not support the `MoveResourceState` RPC. To support this we'd need to:

- Migrate `aws_iam_role_policy` to Terraform Plugin Framework
- Migrate at least part of the `aws_iam_role` schema to Terraform Plugin Framework (for populating the `SourceSchema`)
- Implement a `StateMover` between `aws_iam_role` and `aws_iam_role_policy`

While possible, adding support is unfortunately not as simple as implementing the state movement logic.

As an aside, I'm not sure `moved` is quite correct to model this change. We aren't _moving_ the role resource in the sense that it should no longer be managed, which is the result of a [`moved`](https://developer.hashicorp.com/terraform/language/moved) operation. 

> Before creating a new plan for the resource specified in the `to` field, Terraform checks the state for an existing object at the address specified in the `from` field. Terraform renames existing objects to the string specified in the `to` field and then creates a plan. The plan directs Terraform to provision the resource specified in the `from` field as the resource specified in the `to` field. As a result, Terraform does not destroy the resource during the Terraform run.

I suspect after a successful apply with a move operation, deletion of the `moved` block and an import of the `aws_iam_role` resource would be required to get the "from" resource address back into state. Importing the existing inline policy to the standalone resource may be a more direct path in this situation.

**Comment by lorengordon** [comment]
>As an aside, I'm not sure moved is quite correct to model this change. We aren't moving the role resource in the sense that it should no longer be managed, which is the result of a [moved](https://developer.hashicorp.com/terraform/language/moved) operation.

I think perhaps a further update to `moved` semantics _could_ be in order, to support this type of shift towards resources that align primarily to a single API action. Previously, many resources would use blocks, like `aws_iam_role` with its `inline_policy`, and basically become a resource that manages multiple resources. The last few years, the providers shifted away from that. We also got some great new refactoring tools like `moved`, `import`, and `removed`. But none of them currently support refactoring away from blocks and towards a separate resource.

Note that I was careful in how I phrased the move request previously...

>It would be amazing if I could write a moved block for the module, ***where the `from` is actually the `inline_policy` block*** on the `aws_iam_role` resource, and the `to` is the new `aws_iam_role_policy` resource.

I do not want to move the `aws_iam_role` resource entirely. Just the `inline_policy` block that it was managing.

**Comment by jar-b** [comment]
Makes sense, I'm on the same page now. This sounds like a `moved` enhancement or a new "copy state" feature which functions similar to `moved` but preserves the source resource as-is.

Working within the existing language constraints on the provider side, the `import` block seems the best current migration option. This unfortunately leaves module authors with a decision between a breaking change and retaining a deprecated argument. For our part, we don't intend to fully remove the `inline_policy` argument in the next major release due to the ubiquity of IAM resources. A relevant section of the original proposal:

> Due to the popularity of the resources in this section, argument deprecations are likely to be "soft" deprecations where removal will not happen for several major releases, or until tooling is available to limit the amount of manual changes required to migrate to the preferred pattern. Despite this long removal window, a soft deprecation is still helpful for maintainers to reference when making best practice recommendations to the community.

Hopefully this helps in weighing the available options.

**Comment by jar-b** [comment]
Also, I've added new items in the issue body for the additional suggested resources. Thanks for those @lorengordon! 👍

**Comment by lorengordon** [comment]
>This sounds like a moved enhancement or a new "copy state" feature which functions similar to moved but preserves the source resource as-is.

Yeah. I'll check core and open a feature/enhancement request, if something doesn't already exist.

**Comment by lorengordon** [comment]
> > This sounds like a moved enhancement or a new "copy state" feature which functions similar to moved but preserves the source resource as-is.
> 
> Yeah. I'll check core and open a feature/enhancement request, if something doesn't already exist.

Alright, issue opened: https://github.com/hashicorp/terraform/issues/35785.

**Comment by jar-b** [comment]
Thanks @lusitania 👍 - opened #39822 and added it to the parent issue above.

**Comment by lorengordon** [comment]
Another candidate for exclusive attachments, I think, would be `aws_eks_access_policy_association`.

**Comment by mbbush** [comment]
`aws_ssoadmin_customer_managed_policy_attachment` would be just as useful to have as `aws_ssoadmin_managed_policy_attachment`

(because sso is cross-account, it differentiates with separate API calls between aws-managed policies that have no account id in their ARN, and customer-managed ones that do)

**Comment by lorengordon** [comment]
An interesting resource to add this exclusive attachment for might be `aws_account_region`. It would be nice to be able to say, "this account must be enabled for these regions and only these regions; any other enabled regions should be considered drift and disabled".

**Comment by yermulnik** [comment]
Given the below configuration:
```hcl
resource "aws_iam_role" "role" {
  count                = local.create_role ? 1 : 0
  name                 = "role"
  description          = "Role"
}

resource "aws_iam_role_policy_attachments_exclusive" "role" {
  count       = local.create_role ? 1 : 0
  role_name   = one(aws_iam_role.role[*].name)
  policy_arns = local.managed_policies_list
}
```
we're getting `DeleteConflict` on first `tf destroy`:
```
Error: deleting IAM Role (role): operation error IAM: DeleteRole, https response error StatusCode: 409, RequestID: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx, DeleteConflict: Cannot delete entity, must detach all policies first.
```
and then after 2nd `tf destroy` deletion is happening successfully.

Is this something that should be handled natively by AWS provider?
Adding explicit `depends_on = [aws_iam_role.role]` into `aws_iam_role_policy_attachments_exclusive` didn't help. Still need to run `tf destroy` twice.

AWS provider version is `5.78.0`

Thanks.

**Comment by jar-b** [comment]
Hey @yermulnik - A similar discussion took place over in #39818. This comment might be specifically helpful: https://github.com/hashicorp/terraform-provider-aws/issues/39818#issuecomment-2427407475. 

TLDR; Destruction of `aws_iam_role_policy_attachments_exclusive` indicates that Terraform should no longer monitor the role to synchronize the configured attachments. It does not remove attachments, as that responsibility is reserved for the `aws_iam_role_policy_attachment` resource (or deletion of the parent role resource).

**Comment by yermulnik** [comment]
@jar-b Thanks for the pointer 👍🏻

**Comment by lorengordon** [comment]
Another resource that would be pretty nice to have an exclusive option would be [`aws_vpc_block_public_access_exclusion`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_block_public_access_exclusion). If there are any exclusions that are not listed, then that should be considered drift and they should be planned for removal.

**Comment by yermulnik** [comment]
> Organizations policies:
> 
> * [x]  [Exclusive management of aws_organizations_policy_attachments #26352](https://github.com/hashicorp/terraform-provider-aws/issues/26352)

@jar-b Apologies for my ignorance/lack of understanding: the checkmark for this closed-as-not-planned item means it's been resolved but not implemented? If yes, might it worth to mark it with something sort of evident like ⚠️ or ❌?

Thank you for keeping the list updated — it's pretty much convenient to track implementation progress ❤️

**Comment by jar-b** [comment]
Thank you for the heads up. The linked issue was inadvertently marked stale and closed by _automation_. I've re-opened it.
